### DISCUSSION TITLE
Can I use wamr in multithreaded environment?

### DISCUSSION THREAD

**Discussion by NinevskiyK** [bytecodealliance/wasm-micro-runtime#3012] [Q&A] [answered]
Hi! I'm trying to run multiple modules in parallel, using C++ threads. What I'm doing is that:

- Initializing wasm runtime in parent proccess (using `wasm_runtime_full_init`)
- Creating many threads, in each one create and run new module (`wasm_runtime_instantiate`, `wasm_runtime_create_exec_env`, `wasm_runtime_call_wasm`)

 The only wrong thing is exception with a message: "thread signal env not inited". If I comment this check out, everything is working. So what I'm doing wrong, why this exception is raised - there's no any documentation on that. 

Also, I've read [this issue](https://github.com/bytecodealliance/wasm-micro-runtime/issues/512), it is proposed to use `os_thread_create` to run multiple threads, just what I need. But unfortunately, I haven't found any documentation/examples on this, maybe someone has? :)

**Comment by wenyongh** [comment, marked as answer]
Hi, for "thread signal env not inited", maybe you can try calling wasm_runtime_init_thread_env in the beginning of the new thread's callback, and wasm_runtime_destroy_thread_env in the end of the callback. Refer to:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/include/wasm_export.h#L719-L738

**Reply by NinevskiyK** [reply]
Thank you, that works fine!
